compress.cc

```diff
@@ -990,8 +961,80 @@ class sstable_compressor_factory_impl : public sstable_compressor_factory, publi
     }
 };
 
+default_sstable_compressor_factory::default_sstable_compressor_factory(config cfg)
+    : _cfg(std::move(cfg))
+    , _impl(std::make_unique<sstable_compressor_factory_impl>(_cfg))
+{
+    for (shard_id i = 0; i < smp::count; ++i) {
+        auto numa_id = _cfg.numa_config[i];
+        _numa_groups.resize(std::max<size_t>(_numa_groups.size(), numa_id + 1));
+        _numa_groups[numa_id].push_back(i);
+    }
+}
+
+default_sstable_compressor_factory::~default_sstable_compressor_factory() {
+}
+
+auto default_sstable_compressor_factory_config::from_db_config(
+    const db::config& cfg,
+    std::span<const unsigned> numa_config) -> self
+{
+    return self {
+        .register_metrics = true,
+        .enable_writing_dictionaries = cfg.sstable_compression_dictionaries_enable_writing,
+        .memory_fraction_starting_at_which_we_stop_writing_dicts = cfg.sstable_compression_dictionaries_memory_budget_fraction,
+        .numa_config{numa_config.begin(), numa_config.end()},
+    };
+}
 
-future<compressor_ptr> sstable_compressor_factory_impl::make_compressor_for_writing(schema_ptr s) {
+std::vector<unsigned> default_sstable_compressor_factory_config::get_default_shard_to_numa_node_mapping() {
+    auto sp = local_engine->smp().shard_to_numa_node_mapping();
+    return std::vector<unsigned>(sp.begin(), sp.end());
+}
+
+unsigned default_sstable_compressor_factory::local_numa_id() {
+    return _cfg.numa_config[this_shard_id()];
+}
+
+shard_id default_sstable_compressor_factory::get_dict_owner(unsigned numa_id, const sha256_type& sha) {
+    auto hash = read_unaligned<uint64_t>(sha.data());
+    const auto& group = _numa_groups[numa_id];
+    SCYLLA_ASSERT(!group.empty());
```

Comment:
maybe use `on_internal_error()` instead?

Reply:
Done (in the revision I just pushed), although I don't see how it's better.

Reply:
In general we recommend using `on_internal_error()` as it limits the bast radius of bugs, to the operation/table affected by them, instead of automatically expanding it to the entire node (or worse -- cluster) by way of crashes.

Reply:
I know that's the reasoning, but I wonder `on_internal_error` is really better than `SCYLLA_ASSERT` for Scylla's development in the long run. For example, one sad thing about `on_internal_error` is that it doesn't give you a coredump with the problem, only a backtrace. So it might stop a bug from crashing a node, but it might also delay the actual fix

Anyway, in this case it doesn't matter, I'm just grumbling.

Reply:
We want to be able to choose when to get a coredump and when not, even if this means some bugs will be harder to find. Not all customers are willing to donate coredumps, or maybe there are time-periods when it is very inconvenient for them.

Reply:
A backtrace is sad for the developer but a coredump is sad for the operator, especially if it's a coordinated bug that crashes all nodes at once.